cpp/src/arrow/flight/client.cc

```diff
@@ -596,198 +427,256 @@ class GrpcStreamReader : public FlightStreamReader {
+    // not be protected with a lock.
+    if (done_writing_) {
+      return Status::OK();
+    }
+    done_writing_ = true;
+    if (!stream_->WritesDone()) {
+      // Error happened, try to close the stream to get more detailed info
+      return internal::ClientDataStream::Finish(MakeFlightError(
+          FlightStatusCode::Internal, "Could not flush pending record batches"));
     }
     return Status::OK();
   }
 
-  Status Begin(const std::shared_ptr<Schema>& schema,
-               const ipc::IpcWriteOptions& options) override {
-    if (batch_writer_) {
-      return Status::Invalid("This writer has already been started.");
+  Status Finish() override {
+    // This may be used concurrently by reader/writer side of a
+    // stream, so it needs to be protected.
+    std::lock_guard<std::mutex> guard(finish_mutex_);
+
+    // Now that we're shared between a reader and writer, we need to
+    // protect ourselves from being called while there's an
+    // outstanding read.
+    std::unique_lock<std::mutex> read_guard(read_mutex_, std::try_to_lock);
+    if (!read_guard.owns_lock()) {
+      return MakeFlightError(FlightStatusCode::Internal,
+                             "Cannot close stream with pending read operation.");
     }
-    std::unique_ptr<ipc::internal::IpcPayloadWriter> payload_writer(
-        new DoPutPayloadWriter<ProtoReadT, FlightReadT>(
-            descriptor_, std::move(rpc_), write_size_limit_bytes_, writer_, this));
-    // XXX: this does not actually write the message to the stream.
-    // See Close().
-    ARROW_ASSIGN_OR_RAISE(batch_writer_, ipc::internal::OpenRecordBatchWriter(
-                                             std::move(payload_writer), schema, options));
-    return Status::OK();
+
+    // Try to flush pending writes. Don't use our WritesDone() to
+    // avoid recursion.
+    bool finished_writes = done_writing_ || stream_->WritesDone();
+    done_writing_ = true;
+
+    Status st = Base::Finish();
+    if (!finished_writes) {
+      return Status::FromDetailAndArgs(
+          st.code(), st.detail(), st.message(),
+          ". Additionally, could not finish writing record batches before closing");
+    }
+    return st;
   }
 
-  Status Begin(const std::shared_ptr<Schema>& schema) override {
-    return Begin(schema, options_);
+  using Base::stream_;
+  std::mutex read_mutex_;
+  std::mutex finish_mutex_;
+  bool done_writing_;
+};
+
+class GrpcClientGetStream
+    : public FinishableDataStream<grpc::ClientReader<pb::FlightData>,
+                                  internal::FlightData> {
+ public:
+  using FinishableDataStream::FinishableDataStream;
+
+  bool ReadData(internal::FlightData* data) override {
+    bool success = internal::ReadPayload(stream_.get(), data);
+    if (ARROW_PREDICT_FALSE(!success)) return false;
+    if (data->body &&
+        ARROW_PREDICT_FALSE(!data->body->device()->Equals(*memory_manager_->device()))) {
```

Comment:
Is the `Equals` check actually required? AFAIR, `ViewOrCopy` should just be a no-op from CPU to CPU (and hopefully also for a given CUDA device onto itself).

Reply:
Ah, fair point. I removed the extra check.